src/rust/engine/process_execution/src/docker.rs

```diff
@@ -435,58 +461,123 @@ impl<'a> super::CommandRunner for CommandRunner<'a> {
 
 #[async_trait]
 impl<'a> CapturedWorkdir for CommandRunner<'a> {
-  type WorkdirToken = String;
+  type WorkdirToken = (String, String);
 
   async fn run_in_workdir<'s, 'c, 'w, 'r>(
     &'s self,
-    context: &'c Context,
+    _context: &'c Context,
     _workdir_path: &'w Path,
-    sandbox_path_in_container: Self::WorkdirToken,
+    (container_id, working_dir): Self::WorkdirToken,
     req: Process,
     _exclusive_spawn: bool,
   ) -> Result<BoxStream<'r, Result<ChildOutput, String>>, String> {
     let docker = self.docker.get().await?;
 
-    let env = req
-      .env
-      .iter()
-      .map(|(key, value)| format!("{key}={value}"))
-      .collect::<Vec<_>>();
+    Command::new(req.argv)
+      .env(req.env)
+      .working_dir(working_dir)
+      .spawn(docker, container_id)
+      .await
+  }
 
-    let working_dir = req
-      .working_directory
-      .map(|relpath| Path::new(&sandbox_path_in_container).join(relpath))
-      .unwrap_or_else(|| Path::new(&sandbox_path_in_container).to_path_buf())
-      .into_os_string()
-      .into_string()
-      .map_err(|s| {
-        format!("Unable to convert working directory due to non UTF-8 characters: {s:?}")
-      })?;
+  async fn prepare_workdir_for_capture(
+    &self,
+    _context: &Context,
+    _workdir_path: &Path,
+    (container_id, working_dir): Self::WorkdirToken,
+    req: &Process,
+  ) -> Result<(), String> {
+    // Docker on Linux will frequently produce root-owned output files in bind mounts, because we
+    // do not assume anything about the users that exist in the image.
+    if matches!(
+      Platform::current()?,
+      Platform::Macos_x86_64 | Platform::Macos_arm64
+    ) {
+      return Ok(());
```

Comment:
Comment above aside - is this just attempting to optimize? Clearly if you can run run chmod in the container, you can do it regardless of Host platform; so the only reason not to do it is if it's expensive.

Reply:
Yes, to optimize. macOS will already be slower due to the virtualization and bind mount overheads, so avoiding unnecessary steps is helpful.

Reply:
Your other comment argues for improving/explaining our understanding of the problem, but this one seemingly argues for defensively doing unnecessary things. I don't want to have to include an explanation in my comment for why we do this unnecessarily on some platforms, and so I would prefer for it to exist in code.

I will expand the comment to ensure that it is clear that we don't have a perfect understanding of all of the factors in play.

Reply:
To my understanding we do not know this is un-necessary! You ran an experiment that shows its un-necessary, but is this happenstance of the current volume mount driver implementation? I have no clue. I know you had looked into the past and future of different volume mount drivers. Maybe you have more knowledge here than you're letting on?

Reply:
Basically, I'm reviewing issues and code and have 0 clue about the details of why X happens. I think we have holes in our understanding, but perhaps not?